File: backend/app/models/analysis.py
"""爆款基因分析结果模型"""
import uuid
from datetime import datetime

from sqlalchemy import String, Integer, Float, DateTime, ForeignKey, func
from sqlalchemy.dialects.postgresql import UUID, ARRAY, JSONB
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.database import Base


class AnalysisResult(Base):
    __tablename__ = "analysis_results"

    id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    article_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("articles.id", ondelete="CASCADE"), nullable=False, unique=True, index=True
    )
    # ── 爆款基因6维度 ──
    hotness_score: Mapped[dict] = mapped_column(JSONB, nullable=False)
    # {"level": "S", "score": 92, "factors": ["..."]}
    emotion_tags: Mapped[list | None] = mapped_column(ARRAY(String), nullable=True)
    # ["好奇-惊喜", "焦虑-紧迫"]
    title_formulas: Mapped[list | None] = mapped_column(JSONB, nullable=True)
    # [{"name": "数字+痛点", "confidence": 0.9}]
    structure_template: Mapped[dict | None] = mapped_column(JSONB, nullable=True)
    # {"type": "list", "sections": [...], "golden_ratio": "..."}
    top_3_genes: Mapped[list | None] = mapped_column(JSONB, nullable=True)
    # [{"rank": 1, "gene": "标题公式", "reason": "..."}]
    platform_fit: Mapped[list | None] = mapped_column(JSONB, nullable=True)
    # [{"platform": "xiaohongshu", "fit_score": 85}]
    interaction_analysis: Mapped[dict | None] = mapped_column(JSONB, nullable=True)
    # 互动分析
    # ── LLM 元信息 ──
    raw_llm_response: Mapped[dict | None] = mapped_column(JSONB, nullable=True)
    llm_model: Mapped[str | None] = mapped_column(String(50), nullable=True)
    llm_tokens_used: Mapped[int | None] = mapped_column(Integer, nullable=True)
    # ── 时间戳 ──
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), server_default=func.now())

    # ── 关系 ──
    article: Mapped["Article"] = relationship("Article", back_populates="analysis")

File: server/routes/analysis.ts
// ABOUTME: Analysis routes - viral gene analysis using LLM
import { Router, type Request, type Response } from 'express';
import { getSupabaseClient } from '../src/storage/database/supabase-client';
import { LLMClient, Config } from 'coze-coding-dev-sdk';

const router = Router();

// POST /api/v1/analysis - analyze a single article
router.post('/', async (req: Request, res: Response) => {
  try {
    const { article_id, user_id } = req.body;
    if (!article_id) {
      res.status(400).json({ error: 'article_id 不能为空', code: 40201 });
      return;
    }
    const client = getSupabaseClient();

    // Fetch article
    const { data: article, error: articleError } = await client.from('article').select('*').eq('id', article_id).single();
    if (articleError || !article) {
      res.status(404).json({ error: '文章不存在', code: 40202 });
      return;
    }

    const llmConfig = new Config();
    const llmClient = new LLMClient(llmConfig);

    const analysisPrompt = `你是一位爆款内容分析专家。请分析以下文章的爆款基因。

## 文章标题
${article.title}

## 文章内容
${article.content}

## 分析要求
请从以下6个维度进行深度分析，并严格按照JSON格式输出：
1. **热度评分(hotness_score)**: level(S/A/B/C/D), score(0-100), factors(关键因素数组)
2. **情绪标签(emotion_tags)**: 2-4个情绪标签，格式为"大类-细分"
3. **标题公式(title_formulas)**: 识别使用的标题模式，包含name和confidence
4. **结构模板(structure_template)**: 文章结构类型和段落分析
5. **Top3爆款基因(top_3_genes)**: 最核心的3个爆款因素
6. **平台适配度(platform_fit)**: 各平台适配分数

输出JSON格式：
{
  "hotness_score": {"level": "S", "score": 92, "factors": ["..."]},
  "emotion_tags": ["好奇-惊喜", "焦虑-紧迫"],
  "title_formulas": [{"name": "数字+痛点", "confidence": 0.9}],
  "structure_template": {"type": "list", "description": "..."},
  "top_3_genes": [{"rank": 1, "gene": "...", "reason": "..."}],
  "platform_fit": [{"platform": "xiaohongshu", "fit_score": 85, "reason": "..."}]
}`;

    let analysisData: Record<string, unknown> = {};
    let rawResponse = '';

    try {
      const llmResponse = await llmClient.invoke(
        [{ role: 'user', content: analysisPrompt }],
        { model: 'doubao-seed-2-0-lite-260215', temperature: 0.3 }
      );
      rawResponse = llmResponse.content;
      const jsonMatch = rawResponse.match(/\{[\s\S]*\}/);
      if (jsonMatch) {
        analysisData = JSON.parse(jsonMatch[0]);
      }
    } catch {
      analysisData = { hotness_score: { level: 'C', score: 50, factors: ['分析结果解析失败'] } };
    }

    // Save analysis result
    const { data: result, error } = await client.from('analysis_result').insert({
      article_id,
      user_id: user_id || article.user_id,
      hotness_score: analysisData.hotness_score || { level: 'C', score: 50, factors: [] },
      emotion_tags: (analysisData.emotion_tags as string[]) || [],
      title_formulas: analysisData.title_formulas || [],
      structure_template: analysisData.structure_template || {},
      top_3_genes: analysisData.top_3_genes || [],
      platform_fit: analysisData.platform_fit || [],
      raw_llm_response: rawResponse,
      llm_model: 'coze-bot',
    }).select().single();

    if (error) {
      res.status(500).json({ error: '保存分析结果失败: ' + error.message, code: 40207 });
      return;
    }

    res.status(201).json({ data: result, message: '分析完成' });
  } catch (err) {
    const msg = err instanceof Error ? err.message : 'Unknown error';
    res.status(500).json({ error: '分析失败: ' + msg, code: 40208 });
  }
});

// GET /api/v1/analysis/:article_id - get analysis result for an article
router.get('/:article_id', async (req: Request, res: Response) => {
  try {
    const client = getSupabaseClient();
    const { data, error } = await client.from('analysis_result').select('*').eq('article_id', req.params.article_id).order('created_at', { ascending: false }).limit(1);

    if (error) {
      res.status(500).json({ error: '查询失败: ' + error.message });
      return;
    }
    if (!data || data.length === 0) {
      res.status(404).json({ error: '暂无分析结果', code: 40209 });
      return;
    }
    res.json({ data: data[0] });
  } catch (err) {
    const msg = err instanceof Error ? err.message : 'Unknown error';
    res.status(500).json({ error: '查询失败: ' + msg });
  }
});

export default router;
